core/arch/arm/plat-sam/drivers/clk/at91_main.c

```diff
@@ -0,0 +1,322 @@
+// SPDX-License-Identifier: GPL-2.0-or-later or BSD-3-Clause
+/*
+ *  Copyright (C) 2013 Boris BREZILLON <b.brezillon@overkiz.com>
+ *  Copyright (C) 2021 Microchip
+ */
+
+#include <io.h>
+#include <kernel/delay.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <types_ext.h>
+
+#include "at91_clk.h"
+
+#define SLOW_CLOCK_FREQ		32768
+#define MAINF_DIV		16
+#define USEC_PER_SEC		1000000L
+#define MAINF_LOOP_MIN_WAIT	(USEC_PER_SEC / SLOW_CLOCK_FREQ)
+
+#define OSC_READY_TIMEOUT_US	1000
+
+#define MOR_KEY_MASK		(0xFF << 16)
+
+#define clk_main_parent_select(s)	(((s) & \
+					(AT91_PMC_MOSCEN | \
+					AT91_PMC_OSCBYPASS)) ? 1 : 0)
+
+/*
+ * Main RC Oscillator
+ */
+
+struct main_rc_osc {
+	unsigned long freq;
+	vaddr_t base;
+};
+
+static bool pmc_main_rc_osc_ready(struct main_rc_osc *osc)
+{
+	uint32_t status = io_read32(osc->base + AT91_PMC_SR);
+
+	return !!(status & AT91_PMC_MOSCRCS);
```

Comment:
Actually, I think this could be a problem to remove the `!!`.
bool type can be on 8 bit only. when masking  `status & AT91_PMC_MOSCRCS (BIT (17))`, GCC can infer that the lowest bits will always be 0 and thus optimize to always return false.

Reply:
@clementleger no. C99 standard section 6.3.1.2  Boolean type: "When any scalar value is converted to _Bool, the result is 0 if the value compares equal to 0; otherwise, the result is 1".

Reply:
@jforissier: Hum ok, my experience was probably with a typedefed bool that was not mapping the _Bool keyword. Anyway, thanks for the information about C99 _Bool. Regarding this code, I still think it was made for that purpose, since other check are using BIT(0) without !! (`pmc_main_osc_ready`). I will remove them since not needed